File: 백준/Silver/15649. N과 M （1）/N과 M （1）.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.StringTokenizer;

public class Main {

    static int N,M;
    static int[] arr;
    static boolean[] visited;

    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        arr = new int[M];
        visited = new boolean[N];
        dfs(0);
        System.out.println(sb);


    }

    static void dfs(int depth){
        if(depth == M){
            for(int val : arr){
                sb.append(val).append(" ");
            }
            sb.append("\n");
            return;
        }
        for(int i=0; i<N;i++){
            if(!visited[i]){
                visited[i] = true;
                arr[depth] = i+1;
                dfs(depth+1);
                visited[i] = false;
            }
        }
    }
}

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.js
const readLine = require("readline");

const rl = readLine.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let inputLine = 0;
let arr = [];
let visited;
let N = 0;
let M = 0;
let answer = [];
rl.on("line", (line) => {
  if (inputLine == 0) {
    [N, M] = line.split(" ").map((el) => parseInt(el));
    visited = Array.from({ length: N }, () => false);
    dfs(0);
    console.log(answer.map((arr) => arr.join(" ")).join("\n"));
    process.exit();
  }
});

function dfs(depth) {
  if (depth == M) {
    answer.push([...arr]);
    return;
  }

  for (let i = 0; i < N; i++) {
    if (visited[i] == true) continue;
    visited[i] = true;
    arr.push(i + 1);
    dfs(depth + 1);
    arr.pop(i + 1);
    visited[i] = false;
  }
}
